### DISCUSSION TITLE
Resolving Scheduled Sales Bug

### DISCUSSION THREAD

**Discussion by octaedro** [woocommerce/woocommerce#43643] [General]
A few weeks ago, I started to work on [a bug related to how scheduled sales](https://github.com/woocommerce/woocommerce/issues/41898) are managed. Specifically, when a sale is set to begin in the near future, the sale price isn't updated at the commencement of the sale period as expected. This problem arises because our system updates the product price with the sale price with a [daily cron](https://github.com/woocommerce/woocommerce/blob/c487c84778b02946ce811f89368f369d282f1d3e/plugins/woocommerce/includes/class-wc-install.php#L733) or after [updating the product](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php#L642). The actual product price depends on the calculated start and end times of the sale, and that doesn't imply any product update.

This issue has become more pronounced with the introduction of our new editor, which contrasts with our legacy system. The new editor allows for setting start and end times with minute precision, whereas the old one only supports day-long settings. This enhancement in precision has inadvertently highlighted the underlying issue.

<img width="750" alt="Screenshot 2024-01-11 at 10 43 12" src="https://github.com/woocommerce/woocommerce/assets/1314156/4d295811-f7d8-4419-b27f-f924caec446b">

**Technical approach**

So let's dive into the technical stuff. In the database we have 3 different prices: [price](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/abstracts/abstract-wc-product.php#L270), [regular_price](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/abstracts/abstract-wc-product.php#L280) and [sale_price](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/abstracts/abstract-wc-product.php#L290). Depending on the sale rules, the price might change from the regular to the sale price. The price will be [saved in the database](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php#L653) after determining its type.

In other words, we verify whether the product is on sale or not, and depending on that, we will save the regular or sale price in the database to use it later.

We currently do have a cron task that updates the [product if it's on sale](https://github.com/woocommerce/woocommerce/blob/c487c84778b02946ce811f89368f369d282f1d3e/plugins/woocommerce/includes/wc-product-functions.php#L437), but it [only runs daily](https://github.com/woocommerce/woocommerce/blob/c487c84778b02946ce811f89368f369d282f1d3e/plugins/woocommerce/includes/class-wc-install.php#L733).

This issue for variable products is even more complex since [the prices are cached](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/data-stores/class-wc-product-variable-data-store-cpt.php#L285). Looking at the `is_on_sale` [method for variable products](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-wc-product-variable.php#L505-L510), it relies on these prices to determine if the product is on sale. So, the variable product would be on sale because the prices are different, not because it accomplishes all the requirements needed to be on sale.

In response to this, my initial solution was a quick fix designed to let users continue using the feature for simple products, buying us time to develop a more permanent solution. Collaborating with other teams, we created a temporary fix (see PR [42700](https://github.com/woocommerce/woocommerce/pull/42700)). However, testing revealed that this solution was not universally effective, indicating a need for a more fundamental approach.

Now we are considering three solutions:

**Option 1: Update on Display**

We could continue using our current method of updating the sale price during product updates when the sale is active, adding checks for sale times, each time the product is displayed.
_Pros_: Less invasive, maintains current price handling.
_Cons_: Acts as a temporary fix without addressing the core issue of storing in the database a dynamically calculated value.

**Option 2: Running the update cron more often**

We can run the update cron more often, e.g.: every 5 minutes. I think that we still have to add some modifications to make it work in this case, but it wouldn't be something hard to apply.
_Pros_: Not very invasive in terms of code implementation.
_Cons_: We would update the price after 5 (or 10 or 60) minutes, so it could feel odd for the user not to see their change visible in the moment the sale starts. It would add another cron that it isn't strictly necessary.

**Option 3: On-Demand Price Calculation**

Alternatively, we might calculate the price each time a product is displayed, considering the sale period, and eliminate the database element for current pricing.
_Pros_: Directly resolves the main issue and refines price handling.
_Cons_: Significantly alters our pricing process, potentially impacting extensions.

I'm keen to hear your perspectives, experiences, or any suggestions on this matter.

Thank you!

**Comment by nerrad** [comment]
Thanks for outlining the problem and some potential solutions. I think it's worthwhile investigating and understanding a few more things:

- Are you able to derive from git blame or git history why the saved to DB approach was used? Understanding the historical context might help surface important things.
- I also think it's important to understand why/how this surfaced. My initial looks indicate the implementation has been around for ~7 years. From https://github.com/woocommerce/woocommerce/issues/41898#issuecomment-1841514169, it seems things **were** working fine at one point? I think it's worth finding out if there was a subtle change to something recent that broke it and I don't think the root cause of the bug has been uncovered yet. Of course, it may end up being that it was never a bug, per say, but (as you pointed out) something that became more apparent because the current implementation of sale price scheduling was not _designed_ for the precision the new product editor is using.
- You mention a con of Option 3 is, "potentially impacting extensions". I think it'd be important to understand what those impacts would be. Things like auditing calls to that the relevant methods across extensions and see how they are using it, or auditing the use of any related filters that extensions might be using. 

One thing to keep in mind with the WP Cron approach is that it isn't a _true_ cron job unless it's been configured in a WP environment to be linked with an actual _cron_. Less trafficked sites will have less precision.

**Reply by octaedro** [reply]
> Are you able to derive from git blame or git history why the saved to DB approach was used? Understanding the historical context might help surface important things.

The method `handle_updated_props` was introduced in [this PR](https://github.com/woocommerce/woocommerce/pull/13164) to verify ‘changes’ for terms, meta, and the post object update to prevent unnecessary updates to the DB.

But as far as I can see, we have used this way to handle sale prices (saving them to DB) for over 13 years.

> From https://github.com/woocommerce/woocommerce/issues/41898#issuecomment-1841514169, it seems things were working fine at one point?

I created a brand new JN site with WC 7.9 and changed the timestamp manually after making a product with a scheduled sale, and the problem was still there.

<img width="1026" alt="Screenshot 2024-01-17 at 17 47 46" src="https://github.com/woocommerce/woocommerce/assets/1314156/2ede9a92-8862-4ee5-8065-71363f56d5d7">

BTW, the approach mentioned by @vedanshujain in [his comment](https://github.com/woocommerce/woocommerce/discussions/43643#discussioncomment-8133739) sounds very promising. I think that it would make sense to give that approach a try. I think that [this PR](https://github.com/woocommerce/woocommerce/pull/43497) he did, makes the cron `wc_scheduled_sales ` reliable again.

**Comment by mattsherman** [comment]
> Cons: We would update the price after 5 (or 10 or 60) minutes, so it could feel odd for the user not to see their change visible in the moment the sale starts. It would add another cron that it isn't strictly necessary.

We shouldn't allow users to set sale prices that won't be active immediately when the sale time starts. It is common for stores to have sales that begin and end at very specific times.

**Reply by nerrad** [reply]
> It is common for stores to have sales that begin and end at very specific times

That's a fair point, and I agree - however, adding the time precision is new to the Product Editor and it is effectively a new feature for scheduled sales. The current feature was not designed for this level of precision. I think this means we'd have to rule out wp-cron options (for updating the price in the DB) as a solution but it also means understanding more completely the implications of changing how scheduled sales are handled.

For example, does switching to an on-demand model mean that the logic updating the price in the DB is dead code? Are there filters/actions that extensions would hook into that execute in any of that dead code chain? What are the typical use-cases for those filters/actions? At another level, does this mean the "price" field in the DB is unnecessary and `WC_Product::get_price` should just be a convenience wrapper for getting the current valid price for the product? What would the implications of completely removing that field from the DB?

**Comment by vedanshujain** [comment]
Hey, just curious, perhaps a viable option would be to schedule an action to run at the sale start time that would call the `wc_scheduled_sales();` method.
